### DISCUSSION TITLE
Weird editor context menu entry text -> x-repl

### DISCUSSION THREAD

**Discussion by Chris2011** [apache/netbeans#4894] [Q&A] [answered]
Since NetBeans 9 with the ability of Java Shell inside JDK 9, there were those actions implemented, but just for mimeType text/x-repl:
* Save To Class
* Interrupt exeution

Those where added with @EditorActionRegistration as you can see here: https://github.com/apache/netbeans/blob/master/java/jshell.support/src/org/netbeans/modules/jshell/editor/GenerateClassAction.java#L47-L53 and here: https://github.com/apache/netbeans/blob/master/java/jshell.support/src/org/netbeans/modules/jshell/editor/StopExecutionAction.java#L33-L38

Unfortunately it is not working well, I dunno why but the actions are added to each editor under the popupMenu -> text -> x-repl -> Save to Class and popupMenu -> text -> x-repl -> Interrupt execution

<img width="342" alt="image" src="https://user-images.githubusercontent.com/795658/199073614-7cdd14a5-041a-4699-b917-026dd3d7b505.png">

as you can see (Inisde a Java file, but happens everywhere), this is not correct so I decided to try to fix it. My fix is not about fixing the EditorActionRegistration but to change the actions to the good old ones with ActionRegistration and ActionReferences. Those actions are coming from @sdedic so maybe you have an opinion here and you think that the fix with the EditorActionRegistration is easy, please go ahead, but for me the fix with old action editor and toolbar registration is better than to have this one since years now.

What do you think? The problem is in combination with EditorActionRegistration and the property popupPath. For Actions added with menuPath and toolbarPath there is not a problem, just for the popupMenu.

**Comment by Chris2011** [comment]
So my question is, there are 2 options here, which do we want to implement?
* Fix the original problem
* Change logic to old ActionRegistartion/ActionReference logic

**Reply by Chris2011** [reply]
I prefer the first one, for sure but that needs more time, for me. If someone else knows more insights of the EditorActionRegistration, why this is not working property, please just give me a hint.

**Comment by Chris2011** [comment]
Also, when I fix those actions with my implementation, there is still an empty context menu entry `text -> x-repl -> ` which comes from this file: https://github.com/apache/netbeans/blob/master/java/jshell.support/src/org/netbeans/modules/jshell/resources/layer.xml#L36-L44

Removing this will also remove the empty entry. At the end, I dunno what was the intention for this empty layer registration.

**Comment by matthiasblaesing** [comment, marked as answer]
Grep through the codebase for "Editors/Popup". In ` platform/editor.mimelookup.impl/arch.xml`  I find this:

> Editor Context Menu Actions
> 
> The editor module clients are expected to register popup menu actions
> (javax.swing.Action classes or names of actions
> (i.e. value of Action.NAME attribute) present in editor kit e.g. "goto-source")
> in the Editors/**mime-path**/Popup layer folder.

And indeed - if you look at the system file system, the registrations for the working files are indeed there:

![image](https://user-images.githubusercontent.com/2179736/202026823-f90219ce-b2f2-4ad6-8cd9-953517538892.png)

In this screenshot is the path `Editors/text/css/Popup` is opened and this matches the Popup menu for CSS files. So my quick diagnostic is, that the paths are just messed up.

**Reply by Chris2011** [reply]
Thx for your info and investigation. I also noticed this "Editors/mimetype/Popup" but I didn't know why this wasn't working. I mean for Editors/mimetype/Toolbar it is working well and for ActionRegistration it works well and I know this is not the correct solution it was just a workaround to fix this behaviour and get rid of this context entry.

So with this information, I can try to debug inside of the editor platform popup code to try to figure it out. Thx :)

**Reply by Chris2011** [reply]
I had more time to look at it and when I search for: `https://github.com/apache/netbeans/search?q=popupPosition` it shows me, that jshell is the only one who uses `@EditorActionRegistration(` so I thought by myself, that this could be broken for other mimetypes e.g. text/css and yes for this it is also broken. So it is not specifically to text/x-repl. And what you showed, I guess it is not implemented via @EditorActionRegistration.

When I expand my layer.xml file I also see those misbehaviour:
<img width="255" alt="image" src="https://user-images.githubusercontent.com/795658/204141218-e44b3b3d-d61d-4c8f-8c91-539637e8f40f.png">


as you can see, in Editor we have `Editors -> text -> x-repl -> Popup -> interupt execution` which seems correct and for all other stuff too. If you see above, there is this node `Editors -> Popup -> text -> css -> jshell-save-to-class` so I guess, that the order of adding those is not correct.

**Reply by Chris2011** [reply]
So yes, as you said the paths are messed up seems correct here and I now understand it better after researching and your tips. So let me check the code of EditorActionRegistration :)

**Reply by Chris2011** [reply]
And it seems that I found the problem: https://github.com/apache/netbeans/blob/master/ide/editor.lib2/src/org/netbeans/modules/editor/lib2/EditorActionRegistrationProcessor.java#L239 lemme check this and I will create a PR soon for this.

**Reply by Chris2011** [reply]
@matthiasblaesing so it seems that I just need to change those lines:

<img width="1036" alt="image" src="https://user-images.githubusercontent.com/795658/204142290-5259f4e2-1535-44f8-ace0-8d42f7b3fa42.png">

Clean build and run seems to fix the problem, does it look correct? If yes I will create a PR. I got the lines from the toolbar, implementation below.

**Reply by matthiasblaesing** [reply]
@Chris2011 this looks like the right analysis and the suggested fix looks sane. Will be easier to review with a PR, but sounds right.
